adafruit_led_animation/animation.py

```diff
@@ -0,0 +1,507 @@
+# THE SOFTWARE.
+"""
+`adafruit_led_animation.animation`
+================================================================================
+
+CircuitPython helper library for LED animations.
+
+
+* Author(s): Roy Hooper, Kattni Rembor
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+* `Adafruit NeoPixels <https://www.adafruit.com/category/168>`_
+* `Adafruit DotStars <https://www.adafruit.com/category/885>`_
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://circuitpython.org/downloads
+
+"""
+
+try:
+    from time import monotonic_ns
+except ImportError:
+    import time
+
+    def monotonic_ns():
+        """
+        Implementation of monotonic_ns for platforms without time.monotonic_ns
+        """
+        return int(time.time() * 1000000000)
+
+import random
+from .color import BLACK, RAINBOW
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_LED_Animation.git"
+
+
+class Animation:
+    """
+    Base class for animations.
+    """
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, peers=None, paused=False):
+        self.pixel_object = pixel_object
+        self._speed_ns = 0
+        self.speed = speed  # sets _speed_ns
+        self._color = color
+        self._next_update = monotonic_ns()
+        self.pixel_object.auto_write = False
+        self.color = color
+        self.peers = peers if peers else []
+        self._paused = paused
+        self._time_left_at_pause = 0
+
+    def animate(self):
+        """
+        Call animate() from your code's main loop.  It will draw the animation draw() at intervals
+        configured by the speed property (set from init).
+
+        :return: True if the animation draw cycle was triggered, otherwise False.
+        """
+        if self._paused:
+            return False
+
+        now = monotonic_ns()
+        if now < self._next_update:
+            return False
+
+        self.draw()
+        if self.peers:
+            for peer in self.peers:
+                peer.draw()
+        self.show()
+        for peer in self.peers:
+            peer.show()
+
+        self._next_update = now + self._speed_ns
+        return True
+
+    def draw(self):
+        """
+        Animation subclasses must implement draw() to render the animation sequence.
+        """
+
+    def show(self):
+        """
+        Displays the updated pixels.  Called during animates with changes.
+        """
+        self.pixel_object.show()
+
+    def freeze(self):
+        """
+        Stops the animation until resumed.
+        """
+        self._paused = True
+        self._time_left_at_pause = max(0, monotonic_ns() - self._next_update)
+
+    def resume(self):
+        """
+        Resumes the animation.
+        """
+        self._next_update = monotonic_ns() + self._time_left_at_pause
+        self._time_left_at_pause = 0
+        self._paused = False
```

Comment:
How about replacing `freeze()` and `resume()` with a `paused` property that can be set or return current state?
```python
# to pause
animation.paused = True
# to check
if animation.paused:
    print("Animation currently paused.")
# to un-pause
animation.paused = False
```

Reply:
The freeze/resume protocol on the Animation children grew out of the AnimationSequence and AnimationGroup.  It would be possible to use a paused property, but it doesn't behave cleanly:

AnimationSequence freezes/resumes the current animation in the sequence.
AnimationGroup freezes/resumes all members of the group.

Querying AnimationSequence.paused would return the sequence paused state so that wouldn't be too bad.

Querying AnimationGroup.paused however would need to return all() or any() on the paused state of all members of the group - as it doesn't itself know if they're paused.

I intentionally made the basic signatures for Animations, AnimationGroup and AnimationSequence identical so you can mix and match them in any order.

Reply:
So is the only potential confusion when querying the current state of an AnimationGroup? If freeze/resume acts on all members, how do you get to a state where they are mixed?

Reply:
Having a mixed state in an AnimationGroup seems to defeat the purpose of the Group. If your end goal is a mixed state, you'd want to handle things separately.

@caternuson What is your reasoning behind suggesting this change? I'm trying to understand the purpose of the suggestion.

Reply:
i guess the more important thing here is what do we want the interface to be for pausing and resuming. a property or two methods? @caternuson @kattni

Reply:
Was thinking it would be more Pythonic. Also, to have a way to get current state. But this could be refactored later.

Reply:
Refactoring it later not the right choice as it would change the protocol people use.... So I'll get this refactored ASAP, as @kattni needs this for her latest guide.

Reply:
I'd say keep as is then. The audioio API has separate functions for play/stop/pause:
https://circuitpython.readthedocs.io/en/latest/shared-bindings/audioio/AudioOut.html
so seems OK.

Reply:
Reverted back to the freeze/resume methods.